### PR TITLE
Rewrite data transformer

### PR DESCRIPTION
**Pull request by Solpatium**
Tracking PR

Before merging:

- [x] Add task to remove cairo_vm_range_check, is_felt_pointer and is_uint256 with removal of data_transformer module
- [x] Add simple examples in docstrings for every serializer
- [x] Add dots at the end of error messages
- [x] Update documentation
- [x] Make sure there is enough tests

### CODE REVIEW

File: starknet_py/cairo/serialization/data_serializers/named_tuple_serializer.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+    def serialize_with_context(
+        self,
+        context: SerializationContext,
+        value: Union[Dict, NamedTuple, TupleDataclass],
+    ) -> Generator[int, None, None]:
+        # We can't use isinstance(value, NamedTuple), because there is no NamedTuple type.
+        context.ensure_valid_type(
+            value,
+            isinstance(value, (dict, TupleDataclass)) or self._is_namedtuple(value),
```
Comment (war-in) [inline starknet_py/cairo/serialization/data_serializers/named_tuple_serializer.py:44]: Shouldn't we use `Dict` here?

File: starknet_py/cairo/serialization/data_serializers/named_tuple_serializer.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+            "dict, NamedTuple or TupleDataclass",
+        )
+
+        # noinspection PyUnresolvedReferences, PyProtectedMember
+        values: Dict = value if isinstance(value, dict) else value._asdict()
```
Comment (war-in) [inline starknet_py/cairo/serialization/data_serializers/named_tuple_serializer.py:49]: and here?

**Review by war-in — COMMENTED** [review]
Two questions. I missed the `add tuple serializers` review, so added it here

File: starknet_py/cairo/serialization/data_serializers/uint256_serializer.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+        self, context: SerializationContext, value: int
+    ) -> Generator[int, None, None]:
+        uint256_range_check(value)
+        result = (value % 2**128, value // 2**128)
+        yield from result
+
+    @staticmethod
+    def _ensure_valid_uint128(value: int, context: DeserializationContext):
+        context.ensure_valid_value(
+            0 <= value < U128_UPPER_BOUND, "expected value in range [0;2^128)"
```
Comment (war-in) [inline starknet_py/cairo/serialization/data_serializers/uint256_serializer.py:46]: Shouldn't we start adding dots at the end of the error messages? (Or is it not the end?)

File: starknet_py/cairo/serialization/data_serializers/named_tuple_serializer.py (hunk 1 above)
Comment (Solpatium) [reply starknet_py/cairo/serialization/data_serializers/named_tuple_serializer.py:44]: No, Dict is the name for the type checking, but dict is the type (https://stackabuse.com/python-check-if-variable-is-a-dictionary/). Fun, right? :D

File: starknet_py/cairo/serialization/data_serializers/uint256_serializer.py (hunk 3 above)
Comment (Solpatium) [reply starknet_py/cairo/serialization/data_serializers/uint256_serializer.py:46]: This was before the error messages discussions. I added it to the list above

File: starknet_py/cairo/serialization/data_serializers/named_tuple_serializer.py (hunk 1 above)
Comment (war-in) [reply starknet_py/cairo/serialization/data_serializers/named_tuple_serializer.py:44]: Yup, very funny 🙄

File: starknet_py/utils/data_transformer/errors.py [hunk 4]
```diff
@@ -1,10 +1,8 @@
...
-class InvalidTypeException(CairoSerializerException, TypeError):
-    """Exception thrown when invalid type was provided"""
-
-
-class InvalidValueException(CairoSerializerException, ValueError):
-    """Exception thrown when invalid value was provided"""
+# This has to be there for backward compatibility.
+# This file will be removed with the whole data_transformer module.
+# pylint: disable=unused-import
+from starknet_py.serialization.errors import (
```
Comment (war-in) [inline starknet_py/utils/data_transformer/errors.py:4]: Please add `TODO: remove with DataTransformer` or something

**Review by war-in — COMMENTED** [review]
We can use Uint256Serializer to serialize returned value from `Account.get_balance`

File: starknet_py/cairo/type_parser.py [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+        """
+        parsed = parse_type(type_string)
+        return self._transform_cairo_lang_type(parsed)
+
+    def _transform_cairo_lang_type(
+        self, cairo_type: cairo_lang_types.CairoType
+    ) -> CairoType:
+        """
+        For now, we use parse function from cairo-lang pacakge. It will be replaced in the future, but we need to hide
+        it from the users.
```
[the remark this answers is not in the document]
Comment (cptartur) [reply starknet_py/cairo/type_parser.py:67]: #684

File: starknet_py/net/udc_deployer/deployer.py [hunk 6]
```diff
@@ -125,3 +122,37 @@ def _compute_address(
             salt=salt,
             deployer_address=deployer_address,
         )
+
+
+deployer_abi = AbiParser(
```
[the remark this answers is not in the document]
Comment (cptartur) [reply starknet_py/net/udc_deployer/deployer.py:127]: I think so, also the `deployer_serializer` below.

File: starknet_py/utils/data_transformer/errors.py (hunk 4 above)
Comment (cptartur) [reply starknet_py/utils/data_transformer/errors.py:4]: #685

**Review by war-in — APPROVED** [review]
Let's merge it 🔥

**Pull request merged**